### ISSUE TITLE
Mixed-precision mode in keras: 'AutoCastVariable' object is not subscriptable

### ISSUE THREAD

**Issue by nsmetanin** [tensorflow/tensorflow#30417]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 LTS
- TensorFlow installed from (source or binary): official Tensorflow docker = tensorflow/tensorflow:1.14.0-gpu-py3
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6.8

**Describe the current behavior**
The following error occurs in a given code below when trying to enable mixed-precision mode with `keras.mixed_precision.experimental.set_policy('infer_float32_vars')`:
```
    auto_cast_variable_slice_repr.py:16 call
        return keras.backend.dot(inputs, w[:16, :16])

    TypeError: 'AutoCastVariable' object is not subscriptable
```

It looks like _ResourceVariable_ (which is returned by self.add_weight with mixed-precision is OFF) supports slice operations, while _AutoCastVariable_ (returned by self.add_weight when mixed-precision is ON) doesn't.

It's possible to workaround this issue by converting this variable into tensor (as shown on line 15), but it's not clear if it's a straightforward way to perform slice op on Variable.

**Describe the expected behavior**
Should work without any error.

**Code to reproduce the issue**
```
import numpy as np

from tensorflow import keras

# Comment this line out to make code complete successfully
keras.mixed_precision.experimental.set_policy('infer_float32_vars')

class MyLayer(keras.layers.Layer):
    def build(self, input_shape):
        self.w = self.add_weight(shape=(16, 16))

    def call(self, inputs, **kwargs):
        w = self.w
        # Uncomment this workaround line below to make it work with mixed-precision ON
        # w = keras.backend.cast(w, dtype=w.dtype)
        return keras.backend.dot(inputs, w[:16, :16])

input = keras.layers.Input(shape=(16, ))
output = MyLayer()(input)
model = keras.models.Model(input, output)

model.predict(np.zeros(shape=(16, 16)))
```

**Comment by rmothukuru** [comment]
Reproduced the Error with TF Version 1.14

**Comment by reedwm** [comment]
Thank you for the bug report. This should be fixed by 604988b5d4e8cec6564db6502e6e40eefac8fc67.
